File: pkg/controller/client/client_reports_plane.go
// Copyright 2020 The Atlas Authors. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package controller_client

import (
	"context"
	log "github.com/sirupsen/logrus"

	"github.com/binarly-io/atlas/pkg/api/atlas"
)

// GetTaskStatus requests status(es) of the task
func GetTaskStatus(ReportsPlaneClient atlas.ReportsPlaneClient, taskUUID *atlas.UUID) *DataExchangeResult {
	log.Infof("Status() - start")
	defer log.Infof("Status() - end")

	//ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	// One object request
	taskAddress := atlas.NewAddress(taskUUID)
	objectRequest := atlas.NewObjectRequest()
	objectRequest.SetAddress(taskAddress)
	// Multi-object request
	request := atlas.NewObjectsRequest()
	request.SetRequestDomain(atlas.DomainTask)
	request.SetResultDomain(atlas.DomainStatus)
	request.Append(objectRequest)
	list, err := ReportsPlaneClient.ObjectsReport(ctx, request)
	// Unify call result
	result := NewDataExchangeResult()
	if len(list.GetObjectStatuses()) > 0 {
		result.Recv.ObjectStatus = atlas.NewObjectStatus().SetStatus(list.GetObjectStatuses()[0].GetStatus())
	}
	result.Error = err

	return result
}

// GetTask requests task
func GetTask(ReportsPlaneClient atlas.ReportsPlaneClient, taskUUID *atlas.UUID) *DataExchangeResult {
	log.Infof("Task() - start")
	defer log.Infof("Task() - end")

	//ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	// One object request
	objectRequest := atlas.NewObjectRequest()
	objectRequest.SetAddress(atlas.NewAddress(taskUUID))
	// Multi-object request
	request := atlas.NewObjectsRequest()
	request.SetRequestDomain(atlas.DomainTask)
	request.SetResultDomain(atlas.DomainTask)
	request.Append(objectRequest)
	// Unify call result
	result := NewDataExchangeResult()
	result.Recv.ObjectsList, result.Error = ReportsPlaneClient.ObjectsReport(ctx, request)

	return result
}

// GetTaskReport requests report(es) of the task
func GetTaskReport(ReportsPlaneClient atlas.ReportsPlaneClient, taskUUID *atlas.UUID) *DataExchangeResult {
	log.Infof("Report() - start")
	defer log.Infof("Report() - end")

	//ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	// One object request
	objectRequest := atlas.NewObjectRequest()
	objectRequest.SetAddress(atlas.NewAddress(taskUUID))
	// Multi-object request
	request := atlas.NewObjectsRequest()
	request.SetRequestDomain(atlas.DomainTask)
	request.SetResultDomain(atlas.DomainReport)
	request.Append(objectRequest)
	// Unify call result
	result := NewDataExchangeResult()
	result.Recv.ObjectsList, result.Error = ReportsPlaneClient.ObjectsReport(ctx, request)

	return result
}

// GetTaskFiles requests file(es) of the task
func GetTaskFiles(ReportsPlaneClient atlas.ReportsPlaneClient, taskUUID *atlas.UUID) *DataExchangeResult {
	log.Infof("Files() - start")
	defer log.Infof("Files() - end")

	//ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	// One object request
	objectRequest := atlas.NewObjectRequest()
	objectRequest.SetAddress(atlas.NewAddress(taskUUID))
	// Multi-object request
	request := atlas.NewObjectsRequest()
	request.SetRequestDomain(atlas.DomainTask)
	request.SetResultDomain(atlas.DomainFile)
	request.Append(objectRequest)
	// Unify call result
	result := NewDataExchangeResult()
	result.Recv.ObjectsList, result.Error = ReportsPlaneClient.ObjectsReport(ctx, request)

	return result
}

File: pkg/controller/client/client_reports_plane.js

const {ObjectRequest} = require('../../api/atlas/object_request_pb');
const {ObjectsRequest} = require('../../api/atlas/objects_request_pb');
const {Address} = require('../../api/atlas/address_pb');
const {Domain} = require('../../api/atlas/domain_pb');
const grpc = {};
grpc.web = require('grpc-web');

// gets status of specified task
export async function GetTaskStatus(ReportsPlaneClient, taskUUID) {
    // One object request
    var taskAddress = new Address();
    taskAddress.setUuid(taskUUID);
    var objectRequest = new ObjectRequest();
    objectRequest.setAddress(taskAddress);
    // Multi-object request
    var requestDomain = new Domain();
    requestDomain.setName("task");
    var resultDomain = new Domain();
    resultDomain.setName("status");

    var request = new ObjectsRequest();
    request.setRequestDomain(requestDomain);
    request.setResultDomain(resultDomain);
    request.addRequests(objectRequest);

    /*
    let call = reportsPlaneClient.objectsReport(
        request,
        {"custom-header-1": "value1"},
        callbackGetTaskStatus,
    );
    console.log("set on function GetTaskStatus");
    call.on('status', function(status) {
        console.log('on status')
        console.log(status)
        //if (status.metadata) {
        //    console.log("Received metadata");
        //    console.log(status.metadata);
        //}
    });
    call.on('data', function(data) {
        console.log('on data')
    });
    var completed = false;
    call.on('end', function() {
        console.log('on end')
        completed = true;
    });
    call.on('error', function(error) {
        console.log('on error')
    });

     */
    try {
        const response = await ReportsPlaneClient.objectsReport(
            request,
            {"custom-header-1": "value1"},
        );
        return response
    } catch(error){
        console.log(error);
        return null
    }
}

// gets report of specified task
export async function GetTaskReport(ReportsPlaneClient, taskUUID) {
    // One object request
    var taskAddress = new Address();
    taskAddress.setUuid(taskUUID);
    var objectRequest = new ObjectRequest();
    objectRequest.setAddress(taskAddress);
    // Multi-object request
    var requestDomain = new Domain();
    requestDomain.setName("task");
    var resultDomain = new Domain();
    resultDomain.setName("report");

    var request = new ObjectsRequest();
    request.setRequestDomain(requestDomain);
    request.setResultDomain(resultDomain);
    request.addRequests(objectRequest);

    try {
        const response = await ReportsPlaneClient.objectsReport(
            request,
            {"custom-header-1": "value1"},
        );
        return response
    } catch(error){
        console.log(error);
        return null
    }
}
